### ISSUE TITLE
Remove the version from the pubspec and find a new process for the CHANGELOG

### ISSUE THREAD

**Issue by natebosch** [dart-lang/linter#4389]
Since we aren't publishing the package anymore the `version` config in the pubspec does not have meaning.

Since we roll into the SDK by SHA instead of version tag, the headers in the CHANGELOG don't have meaning. Overall without a publish there isn't a meaningful way for end users to find the _relevant_ changelog entries for the SDK version they are using.

We should stop populating these values.

@pq - do we have some kind of process for editing the SDK CHANGELOG when we update the SHA configured in DEPS for the SDK?

**Comment by srawlins** [comment]
I would be in favor of sort of backfilling the linter CHANGELOG with SDK versions. So starting now we'd write an entry for `# Dart 3.1.0-dev` or something, and add to that list. And then when it's released, we probably have to do some fiddling to put the most recent CHANGELOG entries in the right spot. But it's meaningful. Maybe that double entry with the SDK's actual CHANGELOG is too much fuss?

**Comment by natebosch** [comment]
> Maybe that double entry with the SDK's actual CHANGELOG is too much fuss?

That also made me think of another option:

Changelog here is structured like:

```md
## Next

* Some update since the last SDK sync.

## <Date of most recent SDK sync> - <SDK hash>

* Some update which is in the SDK changelog now.

## 2023-01-01 - <SDK hash>

* Some update which is also in the SDK changelog already.
```

We might need to start by having to do and validate it manually, but I think it could be automated. I think we'd get a lot more safety out of it. cc @devoncarew - some potential features we could look at implementing in general?

The process we'd follow is:

- On PR validate that an appropriate CHANGELOG entry is added under `## NEXT`, or the feature is not user facing (validation could be automated)
- On SDK DEPS change, move any CHANGLEOG entries under `## NEXT` to the SDK changelog. (file change could be automated in bump_deps?)
- After merging SDK DEPS change, open a PR here to rename `## NEXT` to `## <Current Date> - SDK merge` (could be automated with github actions?)

It sacrifices some utility of the CHANGELOG as a user facing file, but I think we gain some utility in making the automation easier, and enforcing CHANGELOG from the start. There is a risk for diff conflicts in outstanding PRs after each SDK sync, so we may want to be sure we can mitigate that before we try it.

**Comment by devoncarew** [comment]
I believe that the main reason that the pub cli tool is exempted from the auto-roll process is due to the fact that they want the recent changes from pub to find their way to the sdk changelog. So this might be a general purpose problem (or at least, might affect a few tools).

To paraphrase the comments from above:

- we want to make sure to record user-impacting changes here as they land
- when the linter rolls into the sdk repo, we want those new changelog entries to find their way to the right place in the sdk changelog
- we may want semantic checkpoints to be written back to this repo ('all the changes from xxx on are in sdk yyy'). this seems a little harder to automate - I'm not sure what this looks like. PRs updating the changelog? tagging commits w/ sdk versions?

But it all seems worth figuring out. In terms of implementation in the sdk, we could either implement it in our rolling tools (`tools/manage_deps.dart` and `tools/rev_sdk_deps.dart`) or have a separate tool which reconciles the sdk changelog w/ some of the changelogs from the deps.

**Comment by natebosch** [comment]
> * I'm not sure what this looks like. PRs updating the changelog?

Yeah I think this should be feasible with github actions, but I don't know how complex it might be. I would think that an action in the SDK repo could fire for every commit and check whether an edit to `DEPS` changed the entry for this repo. If it finds a change it can open a PR here with the CHANGELOG edit.

**Comment by natebosch** [comment]
Another potential workflow:

- In PRs require a changelog entry (or opt-out) somewhere other than the changelog file, like the commit message or the PR in the first comment.
- bump_deps reads the changelog entries and moves them to the SDK changelog file.
- after changing DEPS, update the changelog in this file
- after publishing a stable SDK we also update the changelog in this repo, collapsing the multiple sync headers into the stable version they first appear in.

```md
## <sync date> - <sdk hash>

* Update already in SDK changelog in github.

## <older sync date> - <sdk hash>

* Multiple revs of DEPS between stable publish.

## 3.0.0

* Update already in SDK changelog for published stable SDK.
```

PRs would not risk merge conflicts when they don't have to edit the changelog file.

**Comment by mit-mit** [comment]
I like the idea of having the changelog reflect the SDK releases!

**Comment by pq** [comment]
We've talked for a long time about potentially moving the linter into `sdk/pkg`. Until recently the cons outweighed the pros but we're now seriously looking into it. Assuming we did that, some things would get a little simpler process wise...

Do we have an established process for CHANGELOG updates for other sdk packages?

**Comment by pq** [comment]
> I like the idea of having the changelog reflect the SDK releases!

I think this is a great idea.  I'm inclined to go ahead and make this change. 

@parlough, @natebosch, @jakemac53, @bwilkerson, @devoncarew,  others: thoughts on what I should do with the existing CHANGELOG entries?  Leave them as is? Modify them with their corresponding SDK versions? Have both?

**Comment by natebosch** [comment]
> thoughts on what I should do with the existing CHANGELOG entries?

I don't see much value in trying to edit existing entries - I'd mainly focus on following a new pattern with new entries. I think it's fine for the CHANGELOG file to have a different header format partway through.

**Comment by pq** [comment]
Some take-aways from chats w/ @kevmoo and @munificent:

1. we should document new stuff that we want to help folks discover and
2. we should document anything that might help folks understand (significantly impacting) changes in behavior

For 1, I take it we should document new lints and lint removals. (Deprecations: maybe not?)

For 2, I think we should be judicious (and a lot more conservative than we've been in the past) and reserve entries for lints that have been fundamentally changed or revamped (but not document fixes). If a change is significant enough to signal that it's safe to adopt we might call that out too.  For example, a fixed false positive is probably not enough but if we're graduating a lint to a state of "doneness" by marking it stable, we might call that out too.

Feedback welcome!

**Comment by bwilkerson** [comment]
> For 1, I take it we should document new lints and lint removals. (Deprecations: maybe not?)

It might be even more important to document deprecations, as it will alert some users to the impending removal, giving them time to try to convince us not to remove it.

**Comment by pq** [comment]
With #4427 and #4430, I think this is complete.

Let's follow-up with new issues for any remaining bits.
